safehttp/form.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"mime/multipart"
+	"strconv"
+)
+
+// Form contains parsed data either from URL field's query parameters or from
+// form parameters, part of the body of POST, PATCH or PUT requests that are not
+// multipart requests. The fields are only available after parsing the form,
+// through getter functions that specify the type.
+type Form struct {
+	values map[string][]string
+	err    error
+}
+
+// MultipartForm extends the Form structure to define a POST, PATCH or PUT
+// request that is multipart. Its fields are only available after parsing the
+// form, through getter functions that specify the type.
+type MultipartForm struct {
+	*Form
+	file map[string][]*multipart.FileHeader
+}
+
+// Int TODO
+func (f *Form) Int(paramName string, defaultValue int) int {
+	if f.err != nil {
+		return defaultValue
+	}
```

Comment:
I followed this article: https://blog.golang.org/errors-are-values in which, when an error occurs, the subsequent operation become no-ops. Should I create a helper function or have a different behaviour?

Reply:
```
x := f.Int("x", 1)
y := f.Int("y", 2)
z := f.Err()
```

What should the values of `x`, `y` and `z` be if reading `x` results in a parsing error? How can we make it unsurprising?

Reply:
Hmm, I'd expect a developer not to care about the value of x and y if an error occurred in parsing and treat those as garbage since something went wrong and that should be handled. No-oping everything after an error occurred is less surprising than continuing parsing imo. I am leaning towards having this behaviour mainly because we always report the first error that occurred. Otherwise, if we keep parsing, the error might get overwritten in the current implementation. If you prefer the later, I'd say to keep a slice with all the errors that occurred.

Reply:
To pitch in, i've seen strange bugs when someone forgets to handle an error and continues with (partly) initialized return values.

Reply:
It basically boils down to this: if an error occurs, should it be recoverable or not? Form values are controlled by the client, and if a client passes incorrect form values, I'd expect that the server should not put extra effort to fix this and just assume the whole form is broken.

Reply:
Plus, it prevents reflective-type of attacks too. I guess it impacts traditional web apps tough (and not SPAs), but someone can always re-insert those values.